### PR TITLE
[BEAM-3973] Adds a parameter to the Cloud Spanner read connector that can disable batch API

### PR DESCRIPTION
**Pull request by mairbek**

### CODE REVIEW

**Comment by mairbek** [comment]
@chamikaramj @iemejia could you please take a look?

**Comment by iemejia** [comment]
@mairbek It seems this feature is worth a JIRA. You can explain there why it is useful to disable the batch API for example. Can you please create it? Also please squash your commits and add the JIRA id to the commit message.

**Comment by chamikaramj** [comment]
Agree with what Ismaël said. Please create a JIRA for this that explains the issue.

**Comment by mairbek** [comment]
Please take look

**Comment by mairbek** [comment]
@vkedia PTAL

**Comment by iemejia** [comment]
retest this please

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java [hunk 1]
```diff
@@ -193,6 +196,52 @@ public void testQuery() throws Exception {
     p.run();
   }
 
+  @Test
+  public void testReadAll() throws Exception {
```
[the remark this answers is not in the document]
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java]: It does call read all later on, but I've made the name more descriptive

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java [hunk 2]
```diff
@@ -193,6 +196,52 @@ public void testQuery() throws Exception {
...
+    DatabaseClient databaseClient =
```
[the remark this answers is not in the document]
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java]: Done

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java [hunk 3]
```diff
@@ -193,6 +196,52 @@ public void testQuery() throws Exception {
...
+              .set("key")
+              .to((long) i)
+              .set("value")
+              .to(RandomUtils.randomAlphaNumeric(100))
+              .build());
+    }
+
+    databaseClient.writeAtLeastOnce(mutations);
+
+    SpannerConfig spannerConfig = SpannerConfig.create()
```
[the remark this answers is not in the document]
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java]: Done

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java [hunk 4]
```diff
@@ -193,6 +196,52 @@ public void testQuery() throws Exception {
...
+
+    PCollectionView<Transaction> tx =
+        p.apply(
+            SpannerIO.createTransaction()
+                .withSpannerConfig(spannerConfig)
+                .withTimestampBound(TimestampBound.strong()));
+
+    PCollection<Struct> allRecords = p.apply(SpannerIO.read()
+        .withSpannerConfig(spannerConfig)
+        .withBatching(false)
```
[the remark this answers is not in the document]
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java]: So the alternative would to catch the root partitionable exception and fall back to naive read. I prefer to keep this transparent flag here, we'd rather fail the pipeline and give the user feedback, than silently run the inefficient query.

[the remark this answers is not in the document]
Comment (iemejia) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java]: I don't know if I was clear. I don't intend to make this implicit or the IO 'smarter', I think the explicit batching approach is ok, I was just wondering if we can detect eagerly if the query is not root partitionable and the user chose the batching to break even before executing the complete thing. Maybe we can do something like that in a different JIRA (as well as the test for it).

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java [hunk 5]
```diff
@@ -193,6 +184,59 @@ public void testQuery() throws Exception {
...
+        .withProjectId(project)
+        .withInstanceId(options.getInstanceId())
+        .withDatabaseId(databaseName);
+  }
+
+  @Test
+  public void testReadAllRecordsInDb() throws Exception {
+    DatabaseClient databaseClient = getDatabaseClient();
+
+    List<Mutation> mutations = new ArrayList<>();
```
Comment (iemejia) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java]: I referred also to this part too, and what I expected was more in the line of a makeTableData method like [BigtableIOTest's one](https://github.com/apache/beam/blob/50a84326581941bc1edf573a0ad2b798ecb0f6a1/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIOTest.java#L995).

**Comment by iemejia** [comment]
@chamikaramj It should be ok now, Can you please take a second look and merge if you agree.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java [hunk 6]
```diff
@@ -329,12 +333,26 @@ public ReadAll withTimestampBound(TimestampBound timestampBound) {
       return toBuilder().setTimestampBound(timestampBound).build();
     }
 
+    /** If true the uses Cloud Spanner batch API. */
```
Comment (chamikaramj) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: Can you clarify in the documentation that batching is the default ? Alternatively, wow about just having a method withoutBatching() that can be used to disable batching ?

**Comment by chamikaramj** [comment]
Please let me know when the comments are addressed.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java (hunk 6 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: Done. I think `without` method will be more confusing.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java]: Ah, unfortunately Cloud Spanner query planner is not open source. I don't think we are able to implement this today.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java (hunk 5 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadIT.java]: Done

**Comment by mairbek** [comment]
@chamikaramj ptal

**Pull request merged**